### DISCUSSION TITLE
eva_to_memory hangs at `cat $file >$storefifo` command (or NetCat)

### DISCUSSION THREAD

**Discussion by rozwad** [PeterPawn/YourFritz#51] [Q&A]
Hi,

I am trying to upload an image (based on release 07.29) to FB6890 LTE v2. However the script just hangs at the command `cat $file >$storefifo` in [line 172](https://github.com/PeterPawn/YourFritz/blob/ce04c8570956a21b36205d45f7d1a2140b329f30/eva_tools/eva_to_memory#L179).

I suspect that the issue is related to a problem with data transfer indirectly through the pipe/FIFO files and NetCat, which opens (successfully) a connection in [line 172](https://github.com/PeterPawn/YourFritz/blob/ce04c8570956a21b36205d45f7d1a2140b329f30/eva_tools/eva_to_memory#L172) with command `nc -w 3 $data_ip $data_port <&$upstream 2>/dev/null 1>&2 &`, however for some reason there is no actual data transfer. Tried to debug and see, if the NetCat command outputs any errors by replacing the command in [line 172](https://github.com/PeterPawn/YourFritz/blob/ce04c8570956a21b36205d45f7d1a2140b329f30/eva_tools/eva_to_memory#L172) with `nc -D -v $data_ip $data_port <&$upstream &`.  However the script also just hangs without any output.

Tried on another, faulty FB6890 LTE v2 (it keeps rebooting every 2-3 minutes) – the image upload is successful (though the box keeps restarting).

Would appreciate any hints, what could be the reason for "hanging" the script when trying to flash the (former) FB6890 LTE v2 or how to troubleshoot that further.

Best regards.

**Comment by PeterPawn** [comment]
The option `-D` seems useless to me in the (changed) call (it enables the socket option(!) DEBUG in my version of `netcat`) - but if the modified line containing the `-v` option does not lead to a message like `Connection to <ip> <port> succeeded!` (on STDERR of the `netcat` invocation), obviously the connection was not established - and it's not a miracle, if the later `cat` hangs in this case.

Try to capture the network traffic and search for the 3-way handshake (by TCP stack) for this connection - there should be a SYN-SYN-ACK sequence for the data connection (don't mix up this one with the control connection, where commands and replies are transmitted) or any hint, why the data connection could not be opened.

And at the same time, you should provide a listing of the control connection - either from the automatically created log file or an extraction of the dialogs from the capture mentioned above. 

The "interpreter" for server responses is only a very dumb/primitive one (in my script - only the (message) numbers are checked) - any unexpected answer may (or better: will) baffle the code.

**Comment by rozwad** [comment]
Thank you indeed for prompt response and the hints.
Let me clarify – the netcat verbose output (for `nc -v`) actually confirms the connection is established:
```
linux:~/repos/YourFritz/eva_tools$ bash ./eva_to_memory ~/repos/freetz-ng/images/6890_07.29.image.in-memory
Found AVM bootloader: AVM EVA Version 1.3292 0x0 0x46409
Found hardware revision: 234
Memory size is 0x0c000000 (192 MB)
Memory size limited to 128 MB
Image size is 0x3185b00 (49 MB)
Setting temporary memory size to: 0x04e7a500
Setting temporary kernel args to: mtdram1=0x84e7a500,0x88000000
Connection to 192.168.178.1 3097 port [tcp/*] succeeded!
^C
linux:~/repos/YourFritz/eva_tools$
```
Please see the fraction of a Wireshark capture (with filter `ip.addr == 192.168.178.1` applied):
```
…	…		…		…		…	…	…
66	5.021844	192.168.178.2	192.168.178.1	TCP	54	63384 → 3097 [ACK] Seq=1 Ack=1 Win=64240 Len=0
67	5.272234	192.168.178.2	192.168.178.1	DNS	84	Standard query 0xe3c8 PTR 3.17.178.52.in-addr.arpa
68	5.364991	192.168.178.2	192.168.178.1	DNS	84	Standard query 0x571e PTR 1.88.108.52.in-addr.arpa
69	6.016713	192.168.178.2	192.168.178.1	FTP	81	Request: STOR 0x84e7a500 0x88000000
70	6.017013	192.168.178.1	192.168.178.2	TCP	60	21 → 63380 [PSH, ACK] Seq=468 Ack=191 Win=46053 Len=0
71	6.017013	192.168.178.1	192.168.178.2	FTP	90	Response: 150 Opening BINARY data connection
72	6.050165	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
73	6.050165	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
74	6.050165	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
75	6.050165	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
76	6.050165	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
77	6.050165	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
78	6.050165	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
79	6.050165	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
80	6.050165	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
81	6.050165	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
82	6.059379	192.168.178.1	192.168.178.2	TCP	60	3097 → 63384 [ACK] Seq=1 Ack=14401 Win=46080 Len=0
83	6.059452	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
…	…		…		…		…		…	…
1442	6.460616	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
1443	6.460616	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
1444	6.460616	192.168.178.2	192.168.178.1	FTP-DATA	1494	[TCP Window Full] FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
1445	6.466566	192.168.178.1	192.168.178.2	TCP	60	3097 → 63384 [ACK] Seq=1 Ack=1716481 Win=46080 Len=0
1446	6.466597	192.168.178.1	192.168.178.2	TCP	60	3097 → 63384 [PSH, ACK] Seq=1 Ack=1720801 Win=44640 Len=0
1447	6.466794	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
1448	6.466794	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
1449	6.466794	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
1450	6.466794	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
1451	6.466794	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
1452	6.466794	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
1453	6.466794	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
1454	6.466794	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
1455	6.466794	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
1456	6.466794	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
1457	6.466794	192.168.178.2	192.168.178.1	FTP-DATA	1494	[TCP Window Full] FTP Data: 1440 bytes (PASV) (STOR 0x84e7a500 0x88000000)
1458	6.781984	192.168.178.2	192.168.178.1	TCP	1494	[TCP Retransmission] 63384 → 3097 [ACK] Seq=1720801 Ack=1 Win=64240 Len=1440
1459	7.288567	192.168.178.2	192.168.178.1	DNS	84	Standard query 0xe3c8 PTR 3.17.178.52.in-addr.arpa
1460	7.381069	192.168.178.2	192.168.178.1	DNS	84	Standard query 0x571e PTR 1.88.108.52.in-addr.arpa
1461	7.395493	192.168.178.2	192.168.178.1	TCP	1494	[TCP Retransmission] 63384 → 3097 [ACK] Seq=1720801 Ack=1 Win=64240 Len=1440
1462	8.024089	192.168.178.2	192.168.178.1	DNS	76	Standard query 0x73bc A dns.msftncsi.com
1463	8.608680	192.168.178.2	192.168.178.1	TCP	1494	[TCP Retransmission] 63384 → 3097 [ACK] Seq=1720801 Ack=1 Win=64240 Len=1440
1464	9.392117	192.168.178.2	192.168.178.1	DNS	84	Standard query 0x571e PTR 1.88.108.52.in-addr.arpa
1465	11.013265	192.168.178.2	192.168.178.1	TCP	1494	[TCP Retransmission] 63384 → 3097 [ACK] Seq=1720801 Ack=1 Win=64240 Len=1440
1466	11.303827	192.168.178.2	192.168.178.1	DNS	84	Standard query 0xe3c8 PTR 3.17.178.52.in-addr.arpa
1467	13.404972	192.168.178.2	192.168.178.1	DNS	84	Standard query 0x571e PTR 1.88.108.52.in-addr.arpa
```
Packet #1457 is the last one with the `STOR 0x84e7a500 0x88000000` command.
Would appreciate any further hints, how to troubleshoot the issue further.

**Comment by PeterPawn** [comment]
Looks like the data transmission - at least the "sending part" - was done already. You may/should count the packets transmitted or extract (from Wireshark) the payload of the FTP data connection to verify, its size is equal to the image size.

At the same time the lines

```
Memory size is 0x0c000000 (192 MB)
Memory size limited to 128 MB
Image size is 0x3185b00 (49 MB)
```

are looking odd - the reported memory size of 192 MB is too small regarding the installed memory of 512 MB (in all 6890 models I know) and it's further limited to 128 MB as needed by other GRX5 models (7580/7590) to preserve the memory for "bootcore" system above the uploaded image in RAM.

The kernel(s) from file `kernel.image` (there are two kernels, one is the mentioned "bootcore" kernel with its own `initramfs` containing all needed commands and only the second one is the "FRITZ!OS" kernel, which needs the SquashFS image from `filesystem.image`) contain hard-coded addresses for unpacking the kernel code and locating the "entry point" to start the kernel code. There should be a script called `unpack-kernel` in your tree from Freetz files (https://github.com/Freetz/freetz/blob/master/tools/unpack-kernel), which can report those values for a given kernel image. You should check (and show) the addresses from the kernel to make sure, they do not interfere with the area, where the image was loaded (0x84e7a500 to 0x88000000).

But with the reported memory size of 192 MB there's a chance, that the "bootcore" kernel is linked to be used from addresses above 0x8c000000 already - in this case you may call the `eva_to_memory` script with `0` as third parameter (2nd is the IP address of the box) to bypass memory limiting (as needed by 7580/7590).

If all hints above do not work, try it simply with an original firmware image by AVM - it seems, you've added > 5 MB to the filesystem (the original 07.29 uses only 43-44 MB) and if the size is a problem (I've no idea, why the other box should work with the same image in this case, but probably the other device reports a different memory size), you may consider a two-step approach to install the final firmware image - in the first step you've to add only your own public key used for firmware signing and afterwards you may update via AVM's GUI using your own signed file.

And please ... show us/me the log file from FTP dialogs (it's `eva_to_memory.log` in the directory, where the script was called from or you may specify its location with environment variable `EVA_LOG`: https://github.com/PeterPawn/YourFritz/blob/ce04c8570956a21b36205d45f7d1a2140b329f30/eva_tools/eva_to_memory#L32) and not only the output to STDERR.

**Reply by PeterPawn** [reply]
Hmm ... about 1,400 packets with 1,440 bytes of data are only a fraction of data to be sent (49 MB - I've noticed the packet numbers too late) - transmission is stopped only to let the receiver process its buffers and the image isn't even halfway completely transmitted. So there are no needs to verify the size of transferred data - the other hints from above are still valid and you should process one or all (at your option).

**Comment by rozwad** [comment]
> And please ... show us/me the log file from FTP dialogs 

Here you go:
```
220 ADAM2 FTP Server ready
USER adam2
331 Password required for adam2
PASS adam2
230 User adam2 successfully logged in
SYST
215 AVM EVA Version 1.3292 0x0 0x46409
TYPE I
200 Type set to BINARY
MEDIA SDRAM
200 Media set to MEDIA_SDRAM
P@SW
227 Entering Passive Mode (192,168,178,1,12,12)
RETR env
150 Opening BINARY data connection
226 Transfer complete
SETENV memsize 0x04e7a500
200 SETENV command successful
SETENV kernel_args_tmp mtdram1=0x84e7a500,0x88000000
200 SETENV command successful
TYPE I
200 Type set to BINARY
MEDIA SDRAM
200 Media set to MEDIA_SDRAM
P@SW
227 Entering Passive Mode (192,168,178,1,12,19)
STOR 0x84e7a500 0x88000000
150 Opening BINARY data connection
```
Thank you also for the other hints – much appreciated :) Will go through them one-by-one and will provide feedback.

**Comment by rozwad** [comment]
> You should check (and show) the addresses from the kernel to make sure, they do not interfere with the area, where the image was loaded (0x84e7a500 to 0x88000000).

From the `unpack-kernel -i <image>` script output I conclude there is no overlap of the address spaces (or am I wrong?):
```
TI_AR7.LoadAddress=0x80500000
TI_AR7.EntryAddress=0x80C8B500
TI_AR7.RecordLength=0x00328682
TI_AR7.RecordChecksum=0x66527F70
WARNING: calculated TI-record checksum (0x7F7D797E) does not match the saved one (0x66527F70)
TI_AR7.LzmaCompressedStreamLength=0x0032866A
TI_AR7.LzmaUncompressedStreamLength=0x00B47803
TI_AR7.LzmaCompressedStreamChecksum=0x443C9ABF
WARNING: calculated LZMA-stream checksum (0x00000000) does not match the saved one (0x443C9ABF)
TI_AR7_2ND.LoadAddress=0x8DFFFFFC
TI_AR7_2ND.EntryAddress=0x8E691770
TI_AR7_2ND.RecordLength=0x002D92FA
TI_AR7_2ND.RecordChecksum=0x5B1FE8CF
WARNING: calculated TI-record checksum (0x71D26D0A) does not match the saved one (0x5B1FE8CF)
TI_AR7_2ND.LzmaCompressedStreamLength=0x002D92E2
TI_AR7_2ND.LzmaUncompressedStreamLength=0x007D72B9
TI_AR7_2ND.LzmaCompressedStreamChecksum=0x2E0B2504
WARNING: calculated LZMA-stream checksum (0x00000000) does not match the saved one (0x2E0B2504)
```
However not sure if the warnings are relevant or not. What could be the reason for them, and might it have any impact on the issue?

> […] you may call the `eva_to_memory` script with `0` as third parameter (2nd is the IP address of the box) to bypass memory limiting (as needed by 7580/7590).

This uploads indeed the entire image (which size is `0x3185b00` i.e. 51927808 bytes) to the box – the Wireshark captures 36061 FTP-DATA packets, 1440 bytes each, except the last one is 32 bytes less (i.e. just 1408 bytes), so there are 36061 packets × 1440 bytes – 32 bytes (last packet) = 51927808 bytes in total.
```
…
40606	17.031160	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x88e7a500 0x8c000000)
40607	17.031160	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x88e7a500 0x8c000000)
40608	17.031160	192.168.178.2	192.168.178.1	FTP-DATA	1494	[TCP Window Full] FTP Data: 1440 bytes (PASV) (STOR 0x88e7a500 0x8c000000)
40609	17.037316	192.168.178.1	192.168.178.2	TCP	60	3097 → 65150 [ACK] Seq=1 Ack=51883201 Win=46080 Len=0
40610	17.037353	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x88e7a500 0x8c000000)
40611	17.037353	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x88e7a500 0x8c000000)
40612	17.037353	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x88e7a500 0x8c000000)
40613	17.037353	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x88e7a500 0x8c000000)
40614	17.037353	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x88e7a500 0x8c000000)
40615	17.037353	192.168.178.2	192.168.178.1	FTP-DATA	1494	FTP Data: 1440 bytes (PASV) (STOR 0x88e7a500 0x8c000000)
40616	17.037353	192.168.178.2	192.168.178.1	FTP-DATA	1462	FTP Data: 1408 bytes (PASV) (STOR 0x88e7a500 0x8c000000)
40617	17.037385	192.168.178.1	192.168.178.2	TCP	60	3097 → 65150 [PSH, ACK] Seq=1 Ack=51888961 Win=44640 Len=0
40618	17.039367	192.168.178.1	192.168.178.2	TCP	60	3097 → 65150 [PSH, ACK] Seq=1 Ack=51904801 Win=44640 Len=0
40619	17.039607	192.168.178.1	192.168.178.2	TCP	60	3097 → 65150 [PSH, ACK] Seq=1 Ack=51922081 Win=44640 Len=0
40620	17.039657	192.168.178.1	192.168.178.2	TCP	60	3097 → 65150 [PSH, ACK] Seq=1 Ack=51927809 Win=44672 Len=0
40621	19.399385	192.168.178.2	192.168.178.1	DNS	84	Standard query 0xabb9 PTR 2.175.98.52.in-addr.arpa
40622	24.948985	192.168.178.2	192.168.178.1	DNS	83	Standard query 0xad37 A www.msftconnecttest.com
40623	25.128656	192.168.178.2	192.168.178.1	DNS	85	Standard query 0x9a89 PTR 77.92.137.51.in-addr.arpa
40624	25.952170	192.168.178.2	192.168.178.1	DNS	83	Standard query 0xad37 A www.msftconnecttest.com
…
```
Unfortunately after the upload there is no FTP response with `226 Transfer complete`. The `eva_to_memory` script seems to hang at the `cat $file >$storefifo` command. The box does not reboot by itself of course and after power cycle it boots with the old image.

The same symptoms are when uploading with `eva_to_memory` script the in-memory image extracted from original firmware i.e. [FRITZ.Box_6890_LTE-07.29.image](http://ftp.avm.de/fritzbox/fritzbox-6890-lte/other/fritz.os/FRITZ.Box_6890_LTE-07.29.image). The only difference (as compared to the freetz-ng image) is that the script uploads the entire image even without the third parameter (set to `0`, nor the IP address as the second parameter).

Any other ideas how to troubleshoot further?
Eventually would appreciate any hints where I could find any guidance regarding the firmware signing (as this is a new topic to me, shame ;-)

**Comment by rozwad** [comment]
Well, just before having some break, made yet another attempt to upload the image with `eva_to_memory` script. Everything went exactly the same way i.e. had to run the script with the third parameter `0` to bypass the memory limit, the script seemed to "hang" after transferring all FTP-DATA packets, and there was no FTP response with `226 Transfer complete`. I have just killed in shell terminal the leftover `nc` processes (did that also before, after each unsuccessful upload). Then the box rebooted and started with the new image loaded. No idea what happened this time.

Nevertheless, let me thank you very much @PeterPawn for your crucial help.

I do not like to leave such unresolved problems (the root cause for the issue is still unknown). However I have to move on to the other 'faulty' 6890 and try to diagnose, why it keeps restarting every 2-3 minutes. Guess the motherboard is broken. Notwithstanding – I know this not the right thread to discuss that – if by chance anybody has any ideas, where I could start digging from, would appreciate any hints or contact @ priv.
